# Proyecto 1
## Proyecto SIMARGL


Carlos Chew      - 17507
Josue Valenzuela - 171001

### Análisis exploratorio

In [1]:
import pandas as pd

Vamos a usar la parte 1 y 2 del dataset

In [2]:
df1 = pd.read_csv('datasets/dataset-part1.csv')
df1.head(5)

BIFLOW_DIRECTION  DIRECTION DST_TO_SRC_SECOND_BYTES  FIREWALL_EVENT  \
0                 1          0                      40               0   
1                 1          0                       ,               0   
2                 1          0                     104               0   
3                 1          0                       ,               0   
4                 1          0                      40               0   

   FIRST_SWITCHED  FLOW_ACTIVE_TIMEOUT  FLOW_DURATION_MICROSECONDS  \
0      1616660040                  120                         339   
1      1616660040                  120                           0   
2      1616660040                  120                       44725   
3      1616660040                  120                           0   
4      1616660040                  120                        1114   

   FLOW_DURATION_MILLISECONDS  FLOW_END_MILLISECONDS  FLOW_END_SEC  ...  \
0                           0          1616660040010    1616660040  ...   
1                           0          1616660040068    1616660040  ...   
2                          44          1616660040114    1616660040  ...   
3                           0          1616660040122    1616660040  ...   
4                           1          1616660040184    1616660040  ...   

   TCP_WIN_MSS_IN  TCP_WIN_MSS_OUT  TCP_WIN_SCALE_IN  TCP_WIN_SCALE_OUT  \
0            1460                0                 0                  0   
1               0                0                 0                  0   
2               0                0                 0                  0   
3            1440                0                 8                  0   
4               0                0                 0                  0   

   SRC_TOS  DST_TOS  L7_PROTO_NAME SAMPLING_INTERVAL TOTAL_FLOWS_EXP  \
0        0        0        Unknown                 1         2293398   
1       40        0           ICMP                 1         2293400   
2        0        0            TLS                 1         2293404   
3       40        0        Unknown                 1         2293407   
4        0        0            TLS                 1         2293409   

         LABEL  
0  Normal flow  
1  Normal flow  
2  Normal flow  
3  Normal flow  
4  Normal flow  

[5 rows x 50 columns]

In [3]:
df2 = pd.read_csv('datasets/dataset-part2.csv')
df2.head(5)

BIFLOW_DIRECTION  DIRECTION  \
0                 1          0   
1                 1          0   
2                 1          0   
3                 1          0   
4                 1          0   

                             DST_TO_SRC_SECOND_BYTES  FIREWALL_EVENT  \
0  ,,48,,,,,,,,,,40,,,,,,,,,,40,,,,,,,,,,40,,,,,,...               0   
1                                        ,,,,,,,,,,,               0   
2                                        ,,,,,,,,,,,               0   
3                                                  ,               0   
4                                                  ,               0   

   FIRST_SWITCHED  FLOW_ACTIVE_TIMEOUT  FLOW_DURATION_MICROSECONDS  \
0      1618174987                  120                   104116914   
1      1618175067                  120                     9529030   
2      1618175067                  120                     9529169   
3      1618175077                  120                           0   
4      1618175077                  120                           0   

   FLOW_DURATION_MILLISECONDS  FLOW_END_MILLISECONDS  FLOW_END_SEC  ...  \
0                      104116          1618175092090    1618175092  ...   
1                        9529          1618175077010    1618175077  ...   
2                        9529          1618175077010    1618175077  ...   
3                           0          1618175077021    1618175077  ...   
4                           0          1618175077069    1618175077  ...   

   TCP_WIN_MSS_IN  TCP_WIN_MSS_OUT  TCP_WIN_SCALE_IN  TCP_WIN_SCALE_OUT  \
0            1460             1460                 0                  0   
1               0                0                 0                  0   
2               0                0                 0                  0   
3               0                0                 0                  0   
4               0                0                 0                  0   

   SRC_TOS  DST_TOS   L7_PROTO_NAME SAMPLING_INTERVAL TOTAL_FLOWS_EXP  \
0        0      164             TLS                 1        77251491   
1        0        0  STUN.Messenger                 1        77251492   
2      192        0            ICMP                 1        77251493   
3       72        0            ICMP                 1        77251494   
4        0        0         Unknown                 1        77251495   

         LABEL  
0  Normal flow  
1  Normal flow  
2  Normal flow  
3  Normal flow  
4  Normal flow  

[5 rows x 50 columns]

Verifiquemos que tengan las mismas columnas y unamos los datasets en uno solo

In [4]:
list(df1.columns) == list(df2.columns)

True

In [5]:
df = pd.concat([df1, df2])
del df1, df2
df.head(5)

BIFLOW_DIRECTION  DIRECTION DST_TO_SRC_SECOND_BYTES  FIREWALL_EVENT  \
0                 1          0                      40               0   
1                 1          0                       ,               0   
2                 1          0                     104               0   
3                 1          0                       ,               0   
4                 1          0                      40               0   

   FIRST_SWITCHED  FLOW_ACTIVE_TIMEOUT  FLOW_DURATION_MICROSECONDS  \
0      1616660040                  120                         339   
1      1616660040                  120                           0   
2      1616660040                  120                       44725   
3      1616660040                  120                           0   
4      1616660040                  120                        1114   

   FLOW_DURATION_MILLISECONDS  FLOW_END_MILLISECONDS  FLOW_END_SEC  ...  \
0                           0          1616660040010    1616660040  ...   
1                           0          1616660040068    1616660040  ...   
2                          44          1616660040114    1616660040  ...   
3                           0          1616660040122    1616660040  ...   
4                           1          1616660040184    1616660040  ...   

   TCP_WIN_MSS_IN  TCP_WIN_MSS_OUT  TCP_WIN_SCALE_IN  TCP_WIN_SCALE_OUT  \
0            1460                0                 0                  0   
1               0                0                 0                  0   
2               0                0                 0                  0   
3            1440                0                 8                  0   
4               0                0                 0                  0   

   SRC_TOS  DST_TOS  L7_PROTO_NAME SAMPLING_INTERVAL TOTAL_FLOWS_EXP  \
0        0        0        Unknown                 1         2293398   
1       40        0           ICMP                 1         2293400   
2        0        0            TLS                 1         2293404   
3       40        0        Unknown                 1         2293407   
4        0        0            TLS                 1         2293409   

         LABEL  
0  Normal flow  
1  Normal flow  
2  Normal flow  
3  Normal flow  
4  Normal flow  

[5 rows x 50 columns]

In [6]:
# Quitamos los valores invalidos
df = df.dropna()

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12207873 entries, 0 to 8637206
Data columns (total 50 columns):
 #   Column                      Dtype 
---  ------                      ----- 
 0   BIFLOW_DIRECTION            int64 
 1   DIRECTION                   int64 
 2   DST_TO_SRC_SECOND_BYTES     object
 3   FIREWALL_EVENT              int64 
 4   FIRST_SWITCHED              int64 
 5   FLOW_ACTIVE_TIMEOUT         int64 
 6   FLOW_DURATION_MICROSECONDS  int64 
 7   FLOW_DURATION_MILLISECONDS  int64 
 8   FLOW_END_MILLISECONDS       int64 
 9   FLOW_END_SEC                int64 
 10  FLOW_ID                     int64 
 11  FLOW_INACTIVE_TIMEOUT       int64 
 12  FLOW_START_MILLISECONDS     int64 
 13  FLOW_START_SEC              int64 
 14  FRAME_LENGTH                int64 
 15  IN_BYTES                    int64 
 16  IN_PKTS                     int64 
 17  IPV4_DST_ADDR               object
 18  IPV4_SRC_ADDR               object
 19  L4_DST_PORT                 int64 
 20  L

### Pre – procesamiento

Veamos que valores tiene la columna LABEL

In [8]:
df['LABEL'].value_counts()

Normal flow                       6570058
SYN Scan - aggressive             2496814
Denial of Service R-U-Dead-Yet    2276947
Denial of Service Slowloris        864054
Name: LABEL, dtype: int64

Los valores de Denial of Service Slowloris son mucho menos que de los demas. Se usara Oversample una vez las observaciones esten codificadas y separadas en train y test

**Codificamos las variables cualitativas relevantes**

In [9]:
def rename(options):
    return lambda option: options[option]

def get_options(row):
    return dict((v, k) for k, v in enumerate(row.value_counts().reset_index()['index']))

In [11]:
label_options = get_options(df['LABEL'])
label_options

{'Normal flow': 0,
 'SYN Scan - aggressive': 1,
 'Denial of Service R-U-Dead-Yet': 2,
 'Denial of Service Slowloris': 3}

In [12]:
df['LABEL'] = df['LABEL'].apply(rename(label_options))
df['LABEL'].value_counts()

0    6570058
1    2496814
2    2276947
3     864054
Name: LABEL, dtype: int64

In [13]:
df['PROTOCOL_MAP'].value_counts()

tcp          8950058
udp          2791088
icmp          466697
ipv6-icmp         28
gre                2
Name: PROTOCOL_MAP, dtype: int64

In [14]:
df['PROTOCOL_MAP'] = df['PROTOCOL_MAP'].apply(rename(get_options(df['PROTOCOL_MAP'])))
df['PROTOCOL_MAP'].value_counts()

0    8950058
1    2791088
2     466697
3         28
4          2
Name: PROTOCOL_MAP, dtype: int64

In [15]:
df['L7_PROTO_NAME'].value_counts()

Unknown            4090802
HTTP               3237616
DNS                1076385
TLS                 859143
ICMP                360551
                    ...   
NFS.HTTP                 1
Radius.NFS               1
OCS                      1
VHUA.Amazon              1
COAP.BitTorrent          1
Name: L7_PROTO_NAME, Length: 457, dtype: int64

In [16]:
df['L7_PROTO_NAME'] = df['L7_PROTO_NAME'].apply(rename(get_options(df['L7_PROTO_NAME'])))
df['L7_PROTO_NAME'].value_counts()

0      4090802
1      3237616
2      1076385
3       859143
4       360551
        ...   
452          1
453          1
454          1
455          1
456          1
Name: L7_PROTO_NAME, Length: 457, dtype: int64

Las variables  DST_TO_SRC_SECOND_BYTES, IPV4_DST_ADDR, IPV4_SRC_ADDR, SRC_TO_DST_SECOND_BYTES no parecen tener informacion relevante, no las utilizaremos en adelante.

In [17]:
df = df.drop([ 'DST_TO_SRC_SECOND_BYTES', 'IPV4_DST_ADDR', 'IPV4_SRC_ADDR', 'SRC_TO_DST_SECOND_BYTES'], axis=1)

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12207873 entries, 0 to 8637206
Data columns (total 46 columns):
 #   Column                      Dtype
---  ------                      -----
 0   BIFLOW_DIRECTION            int64
 1   DIRECTION                   int64
 2   FIREWALL_EVENT              int64
 3   FIRST_SWITCHED              int64
 4   FLOW_ACTIVE_TIMEOUT         int64
 5   FLOW_DURATION_MICROSECONDS  int64
 6   FLOW_DURATION_MILLISECONDS  int64
 7   FLOW_END_MILLISECONDS       int64
 8   FLOW_END_SEC                int64
 9   FLOW_ID                     int64
 10  FLOW_INACTIVE_TIMEOUT       int64
 11  FLOW_START_MILLISECONDS     int64
 12  FLOW_START_SEC              int64
 13  FRAME_LENGTH                int64
 14  IN_BYTES                    int64
 15  IN_PKTS                     int64
 16  L4_DST_PORT                 int64
 17  L4_SRC_PORT                 int64
 18  LAST_SWITCHED               int64
 19  MAX_IP_PKT_LEN              int64
 20  MIN_IP_PKT_LEN         

### Selección de características

In [19]:
from sklearn.feature_selection import SelectKBest, chi2

In [20]:
X = df.drop(['LABEL'], axis=1)
y = df['LABEL']
del df

In [21]:
X.shape

(12207873, 45)

In [22]:
X_new = SelectKBest(chi2, k=10).fit_transform(X, y)

In [23]:
X = X_new
del X_new
X.shape

(12207873, 10)

In [24]:
X = pd.DataFrame(X)

### Separación de datos 
- 55% entrenamiento
- 15% validación
- 30% pruebas

In [25]:
from sklearn.model_selection import train_test_split

In [26]:
feature_train, feature_test, target_train, target_test =\
    train_test_split(X, y, test_size=0.30, random_state=31)

feature_train, feature_validation, target_train, target_validation =\
    train_test_split(feature_train, target_train, test_size=0.21, random_state=31) # 0.21*0.7 = 0.15

#### Resolver el problema del desbalance en las observaciones del ataque Slowloris

In [27]:
from sklearn.utils import resample

In [58]:
X = pd.concat(
    [
        feature_train.reset_index().drop(['index'], axis=1),
        pd.DataFrame(target_train).reset_index().drop(['index'], axis=1)
    ],
    axis=1
)

In [60]:
X.head(5)

0    1              2         3              4      5    6      7  \
0   17901   17  1618232890860  82997240  1618232890842     80  164  64860   
1       0    0  1616673692217   5667065  1616673692217  57782    0      0   
2  899816  899  1618224720923  80257264  1618224720023     80  104    507   
3     742    0  1616669105445   4243088  1616669105444  43560   40   1024   
4     228    0  1616671657493   5027687  1616671657492  57904   40   1024   

       8         9  LABEL  
0  64860  82997240      2  
1      0   5667065      0  
2    507  80257264      2  
3   1024   4243088      1  
4   1024   5027687      1

In [61]:
ft0 = X[X.LABEL == 0]
ft1 = X[X.LABEL == 1]
ft2 = X[X.LABEL == 2]
ft3 = X[X.LABEL == 3]

In [72]:
sizes = [len(ft0), len(ft1), len(ft2), len(ft3)]
sizes

[3633487, 1380346, 1259318, 477802]

In [70]:
max_size = max(sizes)
max_size

3633487

In [71]:
per_size = [x/max_size for x in sizes]
per_size

[1.0, 0.37989567597186946, 0.346586626015175, 0.13149957602710563]

In [75]:
new_size = int(3633487*0.5)
new_size

1816743

Vamos a usar upsample a 1816743 para tener por lo menos un 50% del feature que mas observaciones tiene

In [76]:
ft0 = resample(ft0, replace=True, n_samples=new_size, random_state=27)
ft1 = resample(ft1, replace=True, n_samples=new_size, random_state=27)
ft2 = resample(ft2, replace=True, n_samples=new_size, random_state=27)
ft3 = resample(ft3, replace=True, n_samples=new_size, random_state=27)

In [79]:
X = pd.concat([ft0, ft1, ft2, ft3])
X['LABEL'].value_counts()

0    1816743
1    1816743
2    1816743
3    1816743
Name: LABEL, dtype: int64

In [81]:
feature_train = X.drop(['LABEL'], axis=1)
target_train = X['LABEL']

### Implementación

Algoritmo de árboles de decisión para entrenar el modelo

In [ ]:
from sklearn.multiclass import OneVsRestClassifier
from sklearn import svm
import numpy as np
random_state = np.random.RandomState(0)
dtc = OneVsRestClassifier(
    svm.SVC(kernel="linear", probability=True, random_state=random_state)
)
dtc = dtc.fit(feature_train, target_train)

**Métricas de desempeño del modelo**

Datos de pruebas

In [ ]:
from sklearn.metrics import confusion_matrix, classification_report

In [ ]:
target_pred_test = dtc.predict(feature_test)

In [ ]:
confusion_matrix(target_test, target_pred_test)

In [ ]:
classification_report_test = classification_report(target_test, target_pred_test, target_names=['Normal flow', 'SYN Scan - aggressive', 'Denial of Service R-U-Dead-Yet', 'Denial of Service Slowloris'])
print(classification_report_test)

Datos de validación

In [ ]:
target_pred_validation = dtc.predict(feature_validation)

In [ ]:
confusion_matrix(target_validation, target_pred_validation)

In [ ]:
classification_report_validation = classification_report(target_validation, target_pred_validation, target_names=['Normal flow', 'SYN Scan - aggressive', 'Denial of Service R-U-Dead-Yet', 'Denial of Service Slowloris'])
print(classification_report_validation)

Gráfica de la curva ROC

In [ ]:
from sklearn.metrics import roc_curve, auc

In [ ]:
y_score = dtc.decision_function(feature_test)

In [ ]:
# Compute ROC curve and ROC area for each class
fpr = dict()
tpr = dict()
roc_auc = dict()
for i in range(n_classes):
    fpr[i], tpr[i], _ = roc_curve(y_test[:, i], y_score[:, i])
    roc_auc[i] = auc(fpr[i], tpr[i])

# Compute micro-average ROC curve and ROC area
fpr["micro"], tpr["micro"], _ = roc_curve(y_test.ravel(), y_score.ravel())
roc_auc["micro"] = auc(fpr["micro"], tpr["micro"])

Evaluación cruzada con K - folds para k = 10 